py4j-java/src/main/java/py4j/GatewayServer.java

```diff
@@ -427,6 +427,35 @@ public GatewayServer(Object entryPoint, int port, InetAddress address, int conne
 		this.sSocketFactory = sSocketFactory;
 	}
 
+	/*
+	 * @param gateway gateway instance (or subclass).  Must not be <code>null</code>.
+	 * @param port the host port to usu
+	 * @param address the host address to use
+	 * @param connectTimeout the connect timeout (ms)
+	 * @param readTimeout the read timeout (ms)
+	 * @param customCommands any customCommands to use.  May be <code>null</code>
+	 * @param sSocketFactory socketFactory to use.  Must not be <code>null</code>
+	 */
+	public GatewayServer(Gateway gateway, int port, InetAddress address, int connectTimeout, int readTimeout,
```

Comment:
On second thought, you could keep constructor you created and put the decision logic in the builder (if a gateway field is not null, use the constructor accepting a gateway and ignore the callback client and the entry point fields).